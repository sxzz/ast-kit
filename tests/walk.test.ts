import { describe, expect, expectTypeOf, test, vi } from 'vitest'
import {
  type ExportBinding,
  type ImportBinding,
  babelParse,
  isTypeOf,
  walkASTSetup,
  walkExportDeclaration,
  walkImportDeclaration,
} from '../src'
import type * as t from '@babel/types'

describe('walk', () => {
  test('walkImportDeclaration', () => {
    const ast = babelParse(
      `
      import { a } from 'a' 
      import * as b from 'b'
      import c from 'c'
      import 'd'
      import e, { f } from 'e'
      import type { g } from 'g'
      import type h from 'h'
      import type * as i from 'i'
      import type j from 'j'
      import { type h } from 'h'`,
      'ts',
    )
    const imports: Record<string, ImportBinding> = {}
    for (const stmt of ast.body) {
      if (stmt.type === 'ImportDeclaration')
        walkImportDeclaration(imports, stmt)
    }

    expect(
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      Object.values(imports).map(({ specifier, ...i }) => i),
    ).toMatchInlineSnapshot(`
      [
        {
          "imported": "a",
          "isType": false,
          "local": "a",
          "source": "a",
        },
        {
          "imported": "*",
          "isType": false,
          "local": "b",
          "source": "b",
        },
        {
          "imported": "default",
          "isType": false,
          "local": "c",
          "source": "c",
        },
        {
          "imported": "default",
          "isType": false,
          "local": "e",
          "source": "e",
        },
        {
          "imported": "f",
          "isType": false,
          "local": "f",
          "source": "e",
        },
        {
          "imported": "h",
          "isType": true,
          "local": "h",
          "source": "h",
        },
      ]
    `)
  })

  test('walkASTSetup', async () => {
    const ast = babelParse(
      `
      function foo() {}
      function bar(id: string) {}
      const baz = 1
      `,
      'ts',
    )

    const walkFunctionDeclaration = vi.fn()
    const walkTS = vi.fn()
    const walkIdentifier = vi.fn()

    const p = walkASTSetup(ast, (setup) => {
      setup.onEnter('FunctionDeclaration', (node, parent) => {
        expectTypeOf<t.FunctionDeclaration>(node)
        expectTypeOf<t.ParentMaps['FunctionDeclaration']>(parent)
        walkFunctionDeclaration()
      })
      setup.onEnter(['NumericLiteral', 'TSStringKeyword'], (node, parent) => {
        expectTypeOf<t.NumericLiteral | t.TSStringKeyword>(node)
        expectTypeOf<t.ParentMaps['NumericLiteral' | 'TSStringKeyword']>(parent)
        walkTS()
      })
      setup.onEnter(
        (node): node is t.Identifier => node.type === 'Identifier',
        (node, parent) => {
          expectTypeOf<t.Identifier>(node)
          expectTypeOf<t.ParentMaps['Identifier']>(parent)
          walkIdentifier()
        },
      )

      setup.onEnter((node) => {
        expectTypeOf<t.Node>(node)
        return 123
      })
    })
    expect(walkFunctionDeclaration).toBeCalledTimes(0)

    await p
    expect(walkFunctionDeclaration).toBeCalledTimes(2)
    expect(walkTS).toBeCalledTimes(2)
    expect(walkIdentifier).toBeCalledTimes(4)
  })

  describe('walkExportDeclaration', () => {
    function getExports(code: string) {
      const ast = babelParse(code, 'ts')

      const exports: Record<string, ExportBinding> = {}

      for (const n of ast.body) {
        if (
          !!n &&
          isTypeOf(n, [
            'ExportDefaultDeclaration',
            'ExportAllDeclaration',
            'ExportNamedDeclaration',
          ])
        ) {
          walkExportDeclaration(exports, n)
        }
      }

      return exports
    }

    function mapExports(exports: Record<string, ExportBinding>) {
      return Object.values(exports).map(
        // eslint-disable-next-line unused-imports/no-unused-vars
        ({ declaration, specifier, ...rest }) => rest,
      )
    }

    test.each([
      { code: 'export default function() {}', desc: 'function' },
      { code: 'export default function foo() {}', desc: 'function with id' },
      { code: 'export default class {}', desc: 'class' },
      { code: 'export default class Cls {}', desc: 'class with id' },
      { code: 'export default function tsFoo(): void', desc: 'ts declaration' },
      {
        code: `
          const one = 1
          export default one
        `,
        desc: 'expression with VariableDeclaration',
      },
      { code: 'export default [1]', desc: 'expression' },
    ])('ExportDefaultDeclaration $desc', ({ code }) => {
      const exports = getExports(code)

      expect(mapExports(exports)).matchSnapshot()
    })

    test.each([
      { code: `export * from "a"`, desc: 'value' },
      { code: `export type * from "a"`, desc: 'type' },
    ])('ExportAllDeclaration $desc', ({ code }) => {
      const exports = getExports(code)
      expect(mapExports(exports)).toMatchSnapshot()
    })

    test('ExportNamedDeclaration', () => {
      const exports = getExports(`
        const one = 1
        const two = 2
        export { one, one as aliasOne, type one as specifierTypeOne }
        export type { one as nodeTypeOne }
        export type { two }
  
        export { a, b as aliasB } from "z"
        export * as zAll from "z"
  
        export type { c, d as aliasTypeD } from "z"
        export type * as zTypeAll from "z"
      `)

      expect(mapExports(exports)).toMatchSnapshot()
    })

    /**
     * @see https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#table-export-forms-mapping-to-exportentry-records
     */
    test.each([
      { code: 'export var v', desc: '1' },
      { code: 'export default function f() {}', desc: '2' },
      { code: 'export default function () {}', desc: '3' },
      { code: 'export default 42', desc: '4' },
      { code: `const x = "5";export {x}`, desc: '5' },
      { code: `const v = "v";export {v as x}`, desc: '6' },
      { code: `export {x} from "mod"`, desc: '7' },
      { code: `export {v as x} from "mod"`, desc: '8' },
      { code: `export * from "mod"`, desc: '9' },
      { code: `export * as ns from "mod"`, desc: '10' },
    ])('ExportEntry record fields $desc', ({ code }) => {
      const exports = getExports(code)

      expect(mapExports(exports)).toMatchSnapshot()
    })
  })
})
