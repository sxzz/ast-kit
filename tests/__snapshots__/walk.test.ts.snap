// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`walk > walkExportDeclaration > ExportAllDeclaration 'type' 1`] = `
[
  {
    "exported": "*",
    "isType": true,
    "local": "*",
    "source": "a",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportAllDeclaration 'value' 1`] = `
[
  {
    "exported": "*",
    "isType": false,
    "local": "*",
    "source": "a",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'class with id' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "Cls",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'class' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'expression with VariableDeclaration' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "one",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'expression' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'function with id' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "foo",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'function' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportDefaultDeclaration 'ts declaration' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "tsFoo",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '1' 1`] = `
[
  {
    "exported": "v",
    "isType": false,
    "local": "v",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '2' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "f",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '3' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '4' 1`] = `
[
  {
    "exported": "default",
    "isType": false,
    "local": "default",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '5' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "x",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '6' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "v",
    "source": null,
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '7' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "x",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '8' 1`] = `
[
  {
    "exported": "x",
    "isType": false,
    "local": "v",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '9' 1`] = `
[
  {
    "exported": "*",
    "isType": false,
    "local": "*",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportEntry record fields '10' 1`] = `
[
  {
    "exported": "ns",
    "isType": false,
    "local": "*",
    "source": "mod",
  },
]
`;

exports[`walk > walkExportDeclaration > ExportNamedDeclaration 1`] = `
[
  {
    "exported": "one",
    "isType": false,
    "local": "one",
    "source": null,
  },
  {
    "exported": "aliasOne",
    "isType": false,
    "local": "one",
    "source": null,
  },
  {
    "exported": "specifierTypeOne",
    "isType": true,
    "local": "one",
    "source": null,
  },
  {
    "exported": "nodeTypeOne",
    "isType": true,
    "local": "one",
    "source": null,
  },
  {
    "exported": "two",
    "isType": true,
    "local": "two",
    "source": null,
  },
  {
    "exported": "a",
    "isType": false,
    "local": "a",
    "source": "z",
  },
  {
    "exported": "aliasB",
    "isType": false,
    "local": "b",
    "source": "z",
  },
  {
    "exported": "zAll",
    "isType": false,
    "local": "*",
    "source": "z",
  },
  {
    "exported": "c",
    "isType": true,
    "local": "c",
    "source": "z",
  },
  {
    "exported": "aliasTypeD",
    "isType": true,
    "local": "d",
    "source": "z",
  },
  {
    "exported": "zTypeAll",
    "isType": true,
    "local": "*",
    "source": "z",
  },
]
`;
